lib/rust/lexer/definition/src/lexer.rs

```diff
@@ -227,25 +431,43 @@ impl EnsoLexer {
         }
     }
 
+    /// Triggered when an operator is immediately followed by another (e.g. `.+` or `. ==`).
+    fn on_dotted_operator<R:LazyReader>(&mut self, _reader:&mut R) {
+        trace!(self.logger,"Operator::on_dot_operator");
+        let suffix_check  = self.operator_suffix_check;
+        let offset        = self.offset.consume();
+        let dotted_op     = Token::Operator(".",offset);
+        let current_match = self.consume_current();
+        let second_op_str = current_match.trim_start_matches(|c| c == ' ' || c == '.');
+        let second_offset = current_match.chars().count() - second_op_str.chars().count() - 1;
+        let second_op     = Token::Operator(second_op_str,second_offset);
+        self.append_token(dotted_op);
+        self.append_token(second_op);
+        self.push_state(suffix_check);
+    }
+
     /// The rules for lexing Enso operators.
     fn add_operator_rules(lexer:&mut EnsoLexer) {
-        let operator_char   = Pattern::any_of(Self::operator_chars().as_str());
-        let equals          = c!('=');
-        let comma           = c!(',');
-        let dot             = c!('.');
-        let error_char      = &operator_char | &equals | &comma | &dot;
-        let error_suffix    = &error_char.many1();
-        let operator_body   = &operator_char.many1();
-        let ops_eq          = &equals | l!("==") | l!(">=") | l!("<=") | l!("!=") | l!("#=");
+        let operator_char = Pattern::any_of(Self::operator_chars().as_str());
+        let equals        = c!('=');
+        let comma         = c!(',');
+        let dot           = c!('.');
```

Comment:
are these `c!` and `l!` our macros? To me they are completely unreadable

Reply:
They're shorthand to avoid visual noise in the pattern definitions. As their docs state, `c!(...)` is `Pattern::char` and `l!(...)` is `Pattern::all_of`. They're short for `char` and `literal` respectively.

Reply:
Then either replace them with `char!(...)` and `literal!(...)`, or you can also do `'='.into()` instead of `c!('=')` (which is only 2 chars longer) and you can define Patterns from String to be all of and from char to be char. If I recognize correctly, we have also alias for that in prelude, so you could even write `'='.to()` which will be of equal length to the current solution but without using magical one letter macros